File: app/src/main/java/com/emoji/eatogether/RegisterActivity.java
package com.emoji.eatogether;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.emoji.eatogether.db.User;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RegisterActivity extends AppCompatActivity {
    private FirebaseAuth mAuth;
    private DatabaseReference mUsersRef;
    private Button nextBtn;
    private Button backBtn;
    private final String TAG = "RegisterActivity: ";
    private boolean isFirstStepCompleted = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);
        nextBtn = findViewById(R.id.register_next_and_confirm);
        backBtn = findViewById(R.id.register_cancel_and_back);
        mAuth = FirebaseAuth.getInstance();
        mUsersRef = FirebaseDatabase.getInstance().getReference(User.DB_PREFIX);

        nextBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onClickNext(view);
            }
        });
        backBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onClickCancel(view);
            }
        });
    }

    public void onClickNext(View view) {
        if (!isFirstStepCompleted) {
            onStepForward();
        } else {

        List<Integer> requiredFields = new ArrayList<>(Arrays.asList(
                R.id.register_email_field,
                R.id.register_password_field,
                R.id.register_password_confirmation_field,
                R.id.register_first_name_field,
                R.id.register_last_name_field,
                R.id.register_birthday_field,
                R.id.register_description_field
        ));
        final Map<Integer, String> fields = new HashMap<>();

        boolean isOkay = true;
        for (int i = 0; i < requiredFields.size(); i++) {
            EditText field = findViewById(requiredFields.get(i));
            String value = field.getText().toString();
            fields.put(requiredFields.get(i), value);
            if (value.equals("")) {
                field.setError("Required");
                isOkay = false;
            }
        }
        if (!isOkay) {
            return;
        }

        if (!fields.get(R.id.register_password_field).equals(fields.get(R.id.register_password_confirmation_field))) {
            Toast.makeText(getApplication(), "Password and password confirmation must be the same", Toast.LENGTH_SHORT).show();
            return;
        }
        mAuth.createUserWithEmailAndPassword(fields.get(R.id.register_email_field), fields.get(R.id.register_password_field))
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            // Sign in success, update UI with the signed-in user's information
                            Log.d(TAG, "createUserWithEmail:success");
                            FirebaseUser user = mAuth.getCurrentUser();
                            User userData = new User(mUsersRef.child(user.getUid()));
                            userData.setFirstName(fields.get(R.id.register_first_name_field));
                            userData.setLastName(fields.get(R.id.register_last_name_field));
                            userData.setBirthday(fields.get(R.id.register_birthday_field));
                            userData.setDescription(fields.get(R.id.register_description_field));
                            userData.save();
                            Intent intent = new Intent(getApplication(), ProfileActivity.class);
                            startActivity(intent);
                        } else {
                            // If sign in fails, display a message to the user.
                            Log.d(TAG, "createUserWithEmail:failure", task.getException());
                            Toast.makeText(getApplication(), "Authentication failed.", Toast.LENGTH_SHORT).show();
                        }
                    }
                });
        }
    }

    public void onClickCancel(View view) {
        if (isFirstStepCompleted) {
            onStepBack();
        } else {
            finish();
        }
    }

    private void onStepForward() {
        LinearLayout registerFirstStep = findViewById(R.id.register_first_step);
        registerFirstStep.setVisibility(View.GONE);

        LinearLayout registerSecondStep = findViewById(R.id.register_second_step);
        registerSecondStep.setVisibility(View.VISIBLE);
        isFirstStepCompleted = true;

        nextBtn.setText(R.string.register);
        backBtn.setText(R.string.back);
    }

    private void onStepBack() {
        LinearLayout registerFirstStep = findViewById(R.id.register_first_step);
        registerFirstStep.setVisibility(View.VISIBLE);

        LinearLayout registerSecondStep = findViewById(R.id.register_second_step);
        registerSecondStep.setVisibility(View.GONE);
        isFirstStepCompleted = false;

        nextBtn.setText(R.string.next);
        backBtn.setText(R.string.cancel);
    }

    @Override
    public void onBackPressed() {
        if (isFirstStepCompleted) {
            onStepBack();
        } else {
            super.onBackPressed();
        }
    }
}

File: app/src/main/java/ru/blackbull/eatogether/ui/RegisterActivity.kt
package ru.blackbull.eatogether.ui

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuthInvalidCredentialsException
import com.google.firebase.auth.FirebaseAuthUserCollisionException
import com.google.firebase.auth.FirebaseAuthWeakPasswordException
import kotlinx.android.synthetic.main.activity_register.*
import ru.blackbull.eatogether.R
import ru.blackbull.eatogether.extensions.shortToast
import ru.blackbull.eatogether.models.firebase.User
import ru.blackbull.eatogether.state.RegistrationState
import ru.blackbull.eatogether.ui.viewmodels.FirebaseViewModel
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class RegisterActivity : AppCompatActivity() {
    private var isFirstStepCompleted = false

    private val firebaseViewModel: FirebaseViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        firebaseViewModel.signUpResult.observe(this , Observer { signUpResult ->
            when (signUpResult) {
                is RegistrationState.Error -> {
                    when (signUpResult.exception) {
                        is FirebaseAuthWeakPasswordException -> {
                            shortToast("Weak password")
                        }
                        is FirebaseAuthInvalidCredentialsException -> {
                            shortToast("Email address is malformed")
                        }
                        is FirebaseAuthUserCollisionException -> {
                            shortToast("account with the given email address already exists")
                        }
                    }
                }
                is RegistrationState.Success -> {
                    val intent = Intent(this , ProfileActivity::class.java)
                    startActivity(intent)
                }
                is RegistrationState.Loading -> {
                    shortToast("loading")
                }
            }
        })

        register_next_and_confirm!!.setOnClickListener { view: View? ->
            onClickNext(view)
        }
        register_cancel_and_back!!.setOnClickListener { view: View? ->
            onClickCancel(view)
        }
    }

    private fun onClickNext(view: View?) {
        if (!isFirstStepCompleted) {
            onStepForward()
        } else {
            val userInfo = User()
            userInfo.email = register_email_field.text.toString()
            userInfo.firstName = register_first_name_field.text.toString()
            userInfo.lastName = register_last_name_field.text.toString()
            userInfo.description = register_description_field.text.toString()

            try {
                val formatter = SimpleDateFormat("yyyy-MM-dd" , Locale.US)
                val date = formatter.parse(register_birthday_field.text.toString())
                userInfo.birthday = Timestamp(date!!)
            } catch (e: ParseException) {
                register_birthday_field.error = "Введена некоректная дата"
                return
            }

            val pwd = register_password_field.text.toString()
            val confirmPwd = register_password_confirmation_field.text.toString()
            if (pwd != confirmPwd) {
                register_password_confirmation_field.error = "Пароли не совпадают"
                return
            }
//            firebaseViewModel.signUpWithEmailAndPassword(userInfo , pwd)
        }
    }

    private fun onClickCancel(view: View?) {
        if (isFirstStepCompleted) {
            onStepBack()
        } else {
            finish()
        }
    }

    private fun onStepForward() {
        register_first_step.visibility = View.GONE
        register_second_step.visibility = View.VISIBLE
        isFirstStepCompleted = true
        register_next_and_confirm!!.setText(R.string.register)
        register_cancel_and_back!!.setText(R.string.back)
    }

    private fun onStepBack() {
        register_first_step.visibility = View.VISIBLE
        register_second_step.visibility = View.GONE
        isFirstStepCompleted = false
        register_next_and_confirm!!.setText(R.string.next)
        register_cancel_and_back!!.setText(R.string.cancel)
    }

    override fun onBackPressed() {
        if (isFirstStepCompleted) {
            onStepBack()
        } else {
            super.onBackPressed()
        }
    }
}